# Hands-on Exercise for  FPM Module

### 1. Exploring properties of the dataset accidents_10k.dat. Read more about it here:  http://fimi.uantwerpen.be/data/accidents.pdf

In [1]:
!head accidents_10k.dat

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 
2 5 7 8 9 10 12 13 14 15 16 17 18 20 22 23 24 25 27 28 29 32 33 34 35 36 37 38 39 
7 10 12 13 14 15 16 17 18 20 25 28 29 30 33 40 41 42 43 44 45 46 47 48 49 50 51 52 
1 5 8 10 12 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 41 43 46 48 49 51 52 53 54 55 56 57 58 59 60 61 
5 8 10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 31 33 36 38 39 41 43 46 56 62 63 64 65 66 67 68 
7 8 10 12 17 18 21 23 24 26 27 28 29 30 33 34 35 36 38 41 43 47 59 63 66 69 70 71 72 73 74 75 76 77 78 79 
1 12 14 15 16 17 18 21 22 23 24 25 27 28 29 30 31 35 38 41 43 44 53 56 57 58 59 60 63 66 80 81 82 83 84 
10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 30 31 33 39 41 43 44 46 49 59 60 62 63 66 82 
1 8 10 12 14 15 16 17 18 21 22 23 24 25 27 29 30 31 38 41 43 53 56 59 61 63 66 68 85 86 87 88 89 
1 8 12 13 14 15 16 17 18 22 24 25 28 30 38 41 42 43 46 49 60 63 64 66 80 82 84 90 91 92 93 94 95 


<span style="color:red">**Question 1a:** </span>. How many items are there in the data?

In [2]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' accidents_10k.dat

310


<span style="color:green">**Answer:** there are 310 total item sets </span>

<span style="color:red">**Question 1b:** </span> How many transactions are present in the data?

In [3]:
!wc -l accidents_10k.dat

10000 accidents_10k.dat


<span style="color:green">**Answer:**there are 10,000 tarnsactions </span> 

<span style="color:red">**Question 1c:** </span>.  What is the length of the smallest transaction?

In [52]:
!awk ' {if ( NF < min || NR==1 ) min=NF} END {print min}' accidents_10k.dat

23


<span style="color:green">**Answer:** Length of the smallest tarnsaction is 23 </span>

<span style="color:red">**Question 1d:** </span>  What is the length of the longest transaction?

In [51]:
!awk '{ if (NF > max) max = NF } END { print max }' accidents_10k.dat

45


<span style="color:green">**Answer:** Length of the longest transaction is 45</span> 

<span style="color:red">**Question 1e:** </span>  What is the size of the search space of frequent itemsets in this data?

In [1]:
2**310

2085924839766513752338888384931203236916703635113918720651407820138886450957656787131798913024

<span style="color:green">**Answer:** The search space for the given database will be 2^310 which is present in the output table</span> 

<span style="color:red">**Question 1f:** </span> 
Assume that you work for the deparment of transportation that collected this data. What benefit do you see in using itemset mining approaches on this data?

<span style="color:green">**Answer:**Being in the department of transportation, following dataset can be used to plan future decisions based on the analysing the historic data. We can calculate the common occurences leading to the accidents and try avoiding those combinations in future </span> 

<span style="color:red">**Question 1g:** </span>  What type of itemsets (frequent, maximial or closed) would you be interested in discovering this dataset? State your reason.

<span style="color:green">**Answer:** I would choose closed itemsets mainly because closed itemset is a subset of frequent set and it is also lossless which means it will cover frequency information of all the items and it will give me key items to focus on while preventing accidents</span> 

<span style="color:red">**Question 1h:** </span>  What minsup threshold would you use and why?

<span style="color:green">**Answer:** I will use 20-25% of the accidents which have frequent reasons for their occurences therefore 20-25% of 10,000 accidents will be around 2000-2500 as a minsup value </span> 

### 2. Generating frequent, maximal and closed itemsets using $\color{red}{\text{Apriori}}$, $\color{red}{\text{ECLAT}}$, and $\color{red}{\text{FPGrowth}}$ algorihtms from the dataset accidents_10k.dat 

<span style="color:red">**Question 2a:** </span> Generate frequent itemsets using Apriori, for minsup = 2000, 3000, and 4000. Which of these minsup thresholds results in a maximum number of frequent itemsets? Which of these minsup thresholds results in a least number of frequent itemsets? Provide a rationale for these observations.

In [2]:
!chmod u+x apriori
!./apriori


usage: ./apriori [options] infile [outfile]
find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eva

In [5]:
!./apriori -ts -s-2000 accidents_10k.dat T_AP_Freq_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [18.88s].
writing T_AP_Freq_S2000.txt ... [851034 set(s)] done [0.10s].


In [6]:
!./apriori -ts -s-3000 accidents_10k.dat T_AP_Freq_S3000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.00s].
building transaction tree ... [24741 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.37s].
writing T_AP_Freq_S3000.txt ... [133799 set(s)] done [0.02s].


In [7]:
!./apriori -ts -s-4000 accidents_10k.dat T_AP_Freq_S4000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.34s].
writing T_AP_Freq_S4000.txt ... [29501 set(s)] done [0.00s].


<span style="color:green">**Answer:**From the above code we see that minsup=2000 results in 851034 frequent sets,minsup=3000 results in 133799 frequent sets and minsup=4000 results in 29501 frequent sets. We see here that no. of frequent itemsets is less in higher minsup value as it is obvious that less no. of accidents will have more same attributes resulting in accidents </span> 

<span style="color:red">**Question 2b:** </span>   Using Apriori, compare the execution time for finding frequent itemsets for minsup = 2000, 3000, and 4000. Which of these minsup thresholds takes the least amount of time? Provide a rationale for this observation.

In [8]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [19.04s].
writing <null> ... [851034 set(s)] done [0.01s].
19 secs  846307 microsecs


In [38]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-3000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.00s].
building transaction tree ... [24741 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.40s].
writing <null> ... [133799 set(s)] done [0.00s].
4 secs  829009 microsecs


In [39]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.27s].
writing <null> ... [29501 set(s)] done [0.00s].
1 secs  604548 microsecs


<span style="color:green">**Answer:** Least execution time for frequent itemsets is the one with minsup=4000 which is 1.60 secs and it is least mainly because its minsup=4000 and at this minsup no of frequent itemsets are minimum </span>

<span style="color:red">**Question 2c:** </span> Using Apriori, find the frequent itemsets for minsup = 2000, 3000, and 4000. Determine the number of itemsets for each size (1 to max length of an itemset). What trends do you see that are common for all three minsup thresholds? What trends do you see that are different? Provide a rationale for these observations.

In [15]:
!awk '{print NF-1}' T_AP_Freq_S2000.txt|sort -n|uniq -c

     49 1
    705 2
   5285 3
  23745 4
  69647 5
 139628 6
 195730 7
 193299 8
 133819 9
  63937 10
  20497 11
   4189 12
    483 13
     21 14


In [16]:
!awk '{print NF-1}' T_AP_Freq_S3000.txt|sort -n|uniq -c

     38 1
    468 2
   2830 3
   9887 4
  21779 5
  31964 6
  32020 7
  21862 8
   9839 9
   2705 10
    387 11
     20 12


In [17]:
!awk '{print NF-1}' T_AP_Freq_S4000.txt|sort -n|uniq -c

     33 1
    319 2
   1492 3
   4043 4
   6926 5
   7751 6
   5626 7
   2546 8
    668 9
     91 10
      6 11


<span style="color:green">**Answer:**From the above we see that in all three minsup values we see number of frequent sets with lower no. of items are less and then increases in mid and again decreases towards the tail (which has max no. of items). This is primarily because formation of prefix tree in the apriori is similar which is less in the begining increases in between and again tapers at the end</span> 


<span style="color:red">**Question 2d:** </span>  Using Apriori with minsup=2000, compare the number of frequent, maximal, and closed itemsets. Which is the largest set and which is the smallest set? Provide a rationale for these observations.

In [20]:
!./apriori -tm -s-2000 accidents_10k.dat T_AP_Maximal_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 14 done [29.61s].
filtering for maximal item sets ... done [0.03s].
writing T_AP_Maximal_S2000.txt ... [12330 set(s)] done [0.02s].


In [22]:
!./apriori -tc -s-2000 accidents_10k.dat T_AP_Closed_S2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.00s].
building transaction tree ... [20250 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 14 done [29.66s].
filtering for closed item sets ... done [0.46s].
writing T_AP_Closed_S2000.txt ... [519902 set(s)] done [0.10s].


<span style="color:green">**Answer:** Frequent itemset has the highest number of itemsets followed by closed itemset which has 519902 and maximal has the least number of items sets. This is primarily becasue maximal is a subset of closed and closed is a subset of frequent itemset </span> 

<span style="color:red">**Question 2e:** </span> For a minsup = 2000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [23]:
!chmod u+x eclat
!./eclat

usage: ./eclat [options] infile [outfile]
find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eval. b

In [29]:
!./eclat -ts -s-2000 accidents_10k.dat T_EC_Freq_S2000.txt
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing T_EC_Freq_S2000.txt ... [851034 set(s)] done [0.37s].
./eclat - find frequent item sets with the eclat algorithm
version 5.20 (2017.05.30)        (c) 2002-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
writing <null> ... [851034 set(s)] done [0.27s].
0 secs  530341 microsecs


In [30]:
!chmod u+x fpgrowth
!./fpgrowth
!./fpgrowth -ts -s-2000 accidents_10k.dat T_FP_Freq_S2000.txt
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

usage: ./fpgrowth [options] infile [outfile]
find frequent item sets with the fpgrowth algorithm
version 6.17 (2017.05.30)        (c) 2004-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate e

<span style="color:green">**Answer:** Fp growth takes the least amount of time amongst the three mainly because in FP growth tree is constructed with all necessary support values without scanning through the entire database. This saves lot of time and reduces the execution time. Instead in apriori the execution time increases exponentionally and in fp growth its linearly  </span> 


<span style="color:red">**Question 2f:** </span> For a minsup = 4000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [33]:
!./apriori -ts -s-4000 accidents_10k.dat T_FP_Freq_S4000.txt
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");
!./fpgrowth -ts -s-4000 accidents_10k.dat T_FP_Freq_S4000.txt
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-4000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");
!./eclat -ts -s-4000 accidents_10k.dat T_EC_Freq_S4000.txt
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-4000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.13s].
writing T_FP_Freq_S4000.txt ... [29501 set(s)] done [0.00s].
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7

<span style="color:green">**Answer:** FP growth has taken least time as discussed above that it downt scans through the entire database however the difference between fp growth and eclat's execution time is very less at minsup=4000</span>

<span style="color:red">**Question 2g:** </span>  For a minsup = 6000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [35]:
!./apriori -ts -s-6000 accidents_10k.dat T_FP_Freq_S6000.txt
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-6000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");
!./fpgrowth -ts -s-6000 accidents_10k.dat T_FP_Freq_S6000.txt
import datetime 
start = datetime.datetime.now()
!./fpgrowth -ts -s-6000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");
!./eclat -ts -s-6000 accidents_10k.dat T_EC_Freq_S6000.txt
import datetime 
start = datetime.datetime.now()
!./eclat -ts -s-6000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print(elapsed.seconds,"secs ",elapsed.microseconds,"microsecs");

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.03s].
writing T_FP_Freq_S6000.txt ... [2254 set(s)] done [0.00s].
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [20 item(s)] done [0.00s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.

<span style="color:green">**Answer:** FP growth has taken least time as discussed above that it downt scans through the entire database however the difference between fp growth and eclat's execution time is very less at minsup=6000</span> 

<span style="color:red">**Question 2h:** </span> Fill the following table based on execution times computed in __2e__, __2f__, and __2g__. State your observations on the relative computational efficiency at different support thresholds. Based on your knowledge of these algorithms, provide the reasons behind your observations.

|   Algorithm                |minsup=2000         |minsup=4000         |minsup=6000         |
|----------------------------|--------------------|--------------------|--------------------|    
|Apriori                     |      19.846307     |    1.613870        |        0.277315    |
|Eclat                       |         0.530341   |    0.305021        |        0.274588    |
|FPGrowth                    |            0.305021|      0.256468      |        0.248274    |

<span style="color:green">**Answer:** Fp Growth has always taken the least time as it doesnot scan the databse again and again, however at minsup=6000 the execution time for apriori has reduced considerably. Infact apriori execution time decreases exponentionally with higher minsup i.e at minsup=2000 difference/gap between their execution time was very huge and it keeps on reducing as misup values increases and at minsup=6000 the difference between their execution time is very close. Also according to the formula time complexity formula Apriori = 2^I * D * I and for eclat and fp growth is  2^l * D</span> 

### 3. Discovering frequent subsequences and substrings

Assume that roads in a Cincinnati are assigned numbers. Participants are enrolled in a transportation study and for every trip they make using their car, the sequence of roads taken are recorded. Trips that involves freeways are excluded. This data is in the file <span style="color:blue">road_seq_data.dat</span>.

<span style="color:red">**Question 3a:** </span>  What 'type' of sequence mining will you perform to determine frequently taken 'paths'? Paths are sequences of roads traveresed consecutively in the same order.

<span style="color:green">**Answer:** Out of three avaialble sequence mining approaches-level-wise approach, Projection-based aprroach and Vertical approach, i would use vertical approach as it is faster than all.As in this approach full database is not scanned and support values are generated by inverted indices  </span>  

<span style="color:red">**Question 3b:** </span> How many sequences are there in this sequence database?

In [11]:
!chmod u+x prefixspan
!./prefixspan

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei


Usage: prefixspan [OPTION]... INFILE

       where [OPTION]...  is a list of zero or more optional arguments
             INFILE(s)    is the name of the input transaction database

Additional arguments (at most one input file may be specified):
       -min_sup [minimum support]
       -max_pat [maximum pattern]



In [16]:
!wc -l road_seq_data.dat

1000 road_seq_data.dat


<span style="color:green">**Answer:**Number of sequences in the database is nothing but the total number of rows present in the database which is 1000 in the present database of sequences </span> 

<span style="color:red">**Question 3c:** </span> What is the size of the alphabet in this sequence database?

In [29]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' road_seq_data.dat

1283


<span style="color:green">**Answer:** Alphabet according to the defintiton is the total number of unique elements present in our database. So in this case it is 1283 </span> 

<span style="color:red">**Question 3d:** </span> What are the total number of possible subsequences of length 2 in this dataset?

In [35]:
1283*1283

1646089

<span style="color:green">**Answer:** Here total number of elements avaiable is 1283 and at level 2 which is total possible number of subsequnces of length 2 which will be 1283*1283 = 1646089</span> 

<span style="color:red">**Question 3e:** </span> What are the total number of possible substrings of length 2 in this dataset?

In [37]:
1283*1283

1646089

<span style="color:green">**Answer:** In case of total possible number of substrings we will have to consider 1283*1283 elements</span> 

<span style="color:red">**Question 3f:** </span> Discover frequent __subsequences__ with minsup = 10 and report the number of subsequences discovered.

In [39]:
!./prefixspan -min_sup 10 road_seq_data.dat| sed -n 'p;n'> subseq_activities_data_minsup_10

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei



In [42]:
!wc -l subseq_activities_data_minsup_10

4589 subseq_activities_data_minsup_10


<span style="color:green">**Answer:** Total number of frequent subsequences with min sup 10 is 4589 </span> 

<span style="color:red">**Question 3g:** </span>  Discover frequent __substrings__ with minsup = 10 and report the number of substrings discovered.

In [43]:
!chmod u+x seqwog
!./seqwog

usage: ./seqwog [options] infile [outfile]
find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal sequences, r: rules)
         (target type 'r' implies -a (all occurrences))
-m#      minimum number of items per sequence     (default: 1)
-n#      maximum number of items per sequence     (default: no limit)
-s#      minimum support of a sequence            (default: 10%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of a     rule         (default: 80%)
-a       count all occurrences of a pattern       (default: #sequences)
-F#:#..  support border for filtering item sets   (default: none)
         (list of minimum support values, one per item set size,
         starting at the minimum size, as given with option -m#)
-P#      write

In [46]:
!./seqwog -ts -s-10 road_seq_data.dat substring_result_10

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.01s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [844/1000 transaction(s)] done [0.00s].
writing substring_result_10 ... [613 sequence(s)] done [0.00s].


In [48]:
!wc -l substring_result_10

613 substring_result_10


In [49]:
!cat substring_result_10

95201 13 (10)
95201 (10)
77751 (10)
20483 37613 (10)
20483 (10)
3033 (10)
74826 74848 77336 11 (10)
74826 74848 77336 (10)
74826 74848 (10)
74826 (10)
77336 11 (10)
77336 (10)
40602 (10)
40191 (10)
37708 37730 40191 (10)
37708 37730 (10)
37708 (10)
108 130 2618 37611 (10)
108 130 2618 (10)
108 130 (10)
108 (10)
2618 37611 (10)
2618 (10)
37679 (11)
74797 (11)
77184 13 16 (11)
77184 13 (11)
77184 (11)
40560 (11)
79 (11)
2466 37613 37616 (11)
2466 37613 (11)
2466 (11)
2991 37611 37631 (10)
2991 37611 (11)
2991 (11)
77709 11 31 (10)
77709 11 (11)
77709 (11)
40039 (11)
74767 (13)
37641 40560 (11)
37641 (13)
74784 (13)
49 (13)
39298 (13)
74759 77709 11 31 (10)
74759 77709 11 (11)
74759 77709 (11)
74759 (13)
37649 (13)
37666 (13)
41 2991 37611 37631 (10)
41 2991 37611 (11)
41 2991 (11)
41 (13)
66 (13)
37674 (14)
76435 13 (12)
76435 (14)
1717 37613 (12)
1717 (14)
74792 (14)
74 (14)
74949 13 (12)
74949 (14)
37831 (14)
231 37613 (12)
231 (14)
40579 (15)
3010 37613 (13)
3010 (15)
77728 13 (13)
77

<span style="color:green">**Answer:**Here total number of frequent substring is 613 </span> 

<span style="color:red">**Question 3h:** </span> Explain the difference in the number of frequent subsequences and substrings found in __3f__ and __3g__ above.

<span style="color:green">**Answer:** Number of subsequenc of min_sup 10 is 4589 and in the case of substring the value is 613. The number of substring is relatively lesser than numebr of subsequence primarily becasue subsequence considers various and all combination of all elements without changing the order of all elements whereas substring considers adjacent elements i.e the order in which they are present. For eg subsequence will have combination of first element with second, first element with third, first with last and so on, however in substring it is just the first element with second,second with third and so on (basically adjacent elements).Another reason for a lesser number of substring values is substring is a part or subset of subsequence which makes it a lesser count </span> 